### DISCUSSION TITLE
How to express matrix slicing in basilisp?

### DISCUSSION THREAD

**Discussion by dpom** [basilisp-lang/basilisp#1200] [Q&A] [answered]
I have a matrix (vector of vectors). What is the equivalent expression in basilisp for matrix[a:b, c:d]?

**Comment by ikappaki** [comment, marked as answer]
Hi @dpom,

I assume you are referring to slicing a matrix in a library like `numpy` or `pandas`.

In Basilisp, You will need to use the raw indexing equivalent using `slice`s with `aget` for this

For example:

```clojure
> basilisp repl
basilisp.user=> (import [numpy :as np])
nil

basilisp.user=> (def matrix (np/array [[1, 2, 3],
                                       [4, 5, 6],
                                       [7, 8, 9]]))
#'basilisp.user/matrix

basilisp.user=> (aget matrix #py ((slice 1 3) (slice 1 3)))
array([[5, 6],
       [8, 9]])
```

You could simplify repeated slicing with a custom macro as in the example below, that you could adjust as needed for your specific use cases:

```clojure
basilisp.user=> (defmacro mget [m r1 r2 c1 c2]
                  `(aget ~m (python/tuple [(python/slice ~r1 ~r2) (python/slice ~c1 ~c2)])))
#'basilisp.user/mget

basilisp.user=> (mget matrix 1 3 1 3)
array([[5, 6],
       [8, 9]])
```

related: #1094 

I hope this helps!

**Comment by dpom** [comment]
Thank you for your prompt and complete reply.
I really like your idea of porting clojure to python . I created a project to test basilisp and so far I'm very happy with the results.

**Comment by ikappaki** [comment]
> Thank you for your prompt and complete reply. I really like your idea of porting clojure to python . I created a project to test basilisp and so far I'm very happy with the results.

You are welcome! All credit goes to @chrisrink10 for making Basilisp happen 🙌
